Add `ChainstateManager::ProcessTransaction`

jnewbery opened the pull request:
Similarly to how #18698 added `ProcessNewBlock()` and `ProcessNewBlockHeaders()` methods to the `ChainstateManager` class, this PR adds a new `ProcessTransaction()` method. Code outside validation no longer calls `AcceptToMemoryPool()` directly, but calls through the higher-level `ProcessTransaction()` method. Advantages:

- The interface is simplified. Calling code no longer needs to know about the active chainstate or mempool object, since `AcceptToMemoryPool()` can only ever be called for the active chainstate, and that chainstate knows which mempool it's using. We can also remove the `bypass_limits` argument, since that can only be used internally in validation.
- responsibility for calling `CTxMemPool::check()` is removed from the callers, and run automatically by `ChainstateManager` every time `ProcessTransaction()` is called.

jnewbery committed:
[test] Don't set bypass_limits to true in txvalidationcache_tests.cpp

AcceptToMemoryPool() is called for transactions with fees above
minRelayTxFee and with the mempool not full, so setting bypass_limits to
true or false has no impact on the test.

The only way that changing bypass_limits from true to false could change
the result would be to change the outcome to INVALID(TX_MEMPOOL_POLICY).
Since all the ATMP calls in this test result in VALID both before and
after this change, there is no change in behavior.

jnewbery committed:
[test] Don't set bypass_limits to true in txvalidation_tests.cpp

AcceptToMemoryPool() is called for an invalid coinbase transaction, so
setting bypass_limits to true or false has no impact on the test.

The only way that changing bypass_limits from true to false could change
the result would be to change the outcome to INVALID(TX_MEMPOOL_POLICY).
Since the ATMP call in this test results in INVALID(TX_CONSENSUS) both
before and after this change, there is no change in behavior.

jnewbery committed:
[validation] Add CChainState::ProcessTransaction()

This just calls through to AcceptToMemoryPool() internally, and is currently unused.

Also add a new transaction validation failure reason TX_NO_MEMPOOL to
indicate that there is no mempool.

jnewbery committed:
[refactor] Don't call AcceptToMemoryPool() from outside validation.cpp

jnewbery committed:
[validation] Always call mempool.check() after processing a new transaction

CTxMemPool::check() will carry out internal consistency checks 1/n times,
where n is set by the `-checkmempool` configuration option. By default,
mempool consistency checks are disabled entirely on mainnet.

Therefore, this change has no effect on mainnet nodes running with
default configuration. It simply removes the responsibility to trigger
mempool consistency checks from net_processing.

MarcoFalke commented on src/test/util/setup_common.cpp:
```diff
@@ -315,7 +315,7 @@ CMutableTransaction TestChain100Setup::CreateValidMempoolTransaction(CTransactio
     if (submit) {
         LOCK(cs_main);
-        const MempoolAcceptResult result = AcceptToMemoryPool(m_node.chainman->ActiveChainstate(), *m_node.mempool.get(), MakeTransactionRef(mempool_txn), /* bypass_limits */ false);
+        const MempoolAcceptResult result = m_node.chainman->ProcessTransaction(MakeTransactionRef(mempool_txn), /* bypass_limits */ false);
```
wrong

```suggestion
        const MempoolAcceptResult result = m_node.chainman->ProcessTransaction(MakeTransactionRef(mempool_txn), /* test_accept */ false);
```

jnewbery replied:
Fixed!

maflcko reviewed (commented):
Concept ACK

jamesob commented:
Concept ACK, looks like a good change. Will review soon.

jnewbery commented:
Resolved @MarcoFalke's review comment and added a missing lock annotation.

glozow reviewed (commented):
Concept ACK

practicalswift commented:
Concept ACK

glozow reviewed (commented):
Did a first pass-through. I like that this re-delegates the responsibility of calling `mempool.check()` from net_processing to chainstate manager, and see it as the main purpose of this refactoring.

promag reviewed (commented):
Concept ACK.

Light code review ACK 1d20d4330e21ab600747ced872eba372a08086c5.

jnewbery committed:
[validation] Remove comment about AcceptToMemoryPool()

"This logic is not necessary for memory pool transactions, as
AcceptToMemoryPool already refuses previously-known transaction ids
entirely." refers to the logic at
https://github.com/bitcoin/bitcoin/blob/a206b0ea12eb4606b93323268fc81a4f1f952531/src/main.cpp#L484-L486,
which was later removed in commit 450cbb0944cd20a06ce806e6679a1f4c83c50db2.

glozow reviewed (commented):
code review ACK 3bb34f0c71

A few small suggestions if you retouch. The comment touchups sprinkled throughout the commits make the diffs a little noisy.

ryanofsky commented on src/validation.cpp:
```diff
@@ -3430,7 +3430,9 @@ MempoolAcceptResult ChainstateManager::ProcessTransaction(const CTransactionRef&
     }
-    return AcceptToMemoryPool(active_chainstate, *active_chainstate.m_mempool, tx, /* bypass_limits= */ false, test_accept);
+    auto result = AcceptToMemoryPool(active_chainstate, *active_chainstate.m_mempool, tx, /* bypass_limits= */ false, test_accept);
+    active_chainstate.m_mempool->check(active_chainstate.CoinsTip(), active_chainstate.m_chain.Height() + 1);
```
In commit "[validation] Always call mempool.check() after processing a new transaction" (f87e07c6fe321f0fb97703c82c0e4122f800589f)

This change does not appear to be a pure refactoring, so it would be helpful if commit description mentioned something about cases when these new checks will be done and whether there are effects on validation or performance.

jnewbery replied:
For mainnet nodes with default config, there's no behaviour change. I've added the following text to the commit log:

>    CTxMemPool::check() will carry out internal consistency checks 1/n times,
>    where n is set by the `-checkmempool` configuration option. By default,
>    mempool consistency checks are disabled entirely on mainnet.
>    
>    Therefore, this change has no effect on mainnet nodes running with
>    default configuration. It simply removes the responsibility to trigger
>    mempool consistency checks from net_processing.

ryanofsky commented on src/validation.h:
```diff
@@ -994,6 +994,16 @@ class ChainstateManager
+     * @param[in]  bypass_limits   When true, don't enforce mempool fee and capacity limits.
+     * @param[in]  test_accept     When true, run validation checks but don't submit to mempool.
+     */
+    [[nodiscard]] MempoolAcceptResult ProcessTransaction(const CTransactionRef& tx, bool bypass_limits, bool test_accept=false)
```
In commit "[validation] Add CChainState::ProcessTransaction()" (bbc785f31b5d70c702ae37d259c46652e243ddba)

IMO it would be really good to add this function in its final form without the `bypass_limits` parameter, so the later commit removing it can be dropped. This would simplify reviewing the PR, and more importantly avoid unsafe change:

```diff
-    [[nodiscard]] MempoolAcceptResult ProcessTransaction(const CTransactionRef& tx, bool bypass_limits, bool test_accept=false)
+    [[nodiscard]] MempoolAcceptResult ProcessTransaction(const CTransactionRef& tx, bool test_accept=false)
```

Change is unsafe because a `ProcessTransaction(tx, true)` call would still compile and silently change meaning from "bypass limits and don't test" to "don't bypass limits and do test"

Better to avoid this churn and risk, and just add the new method in final form without temporary (bool, bool) params.

jnewbery replied:
Good idea. This was a misguided attempt to make the commits as granular as possible, but you're right that it just adds churn and confusion.

ryanofsky reviewed (approved):
Code review ACK f87e07c6fe321f0fb97703c82c0e4122f800589f with caveat that I'm not sure if there are maybe performance or other implications from adding more mempool checks in the last commit.

All the other changes are straightforward, easy to review, and clean up the code nicely.

As usual, feel free to ignore review suggestions below, which are all minor.

ryanofsky commented on src/validation.h:
```diff
@@ -994,6 +994,16 @@ class ChainstateManager
+     * @param[in]  bypass_limits   When true, don't enforce mempool fee and capacity limits.
+     * @param[in]  test_accept     When true, run validation checks but don't submit to mempool.
+     */
+    [[nodiscard]] MempoolAcceptResult ProcessTransaction(const CTransactionRef& tx, bool bypass_limits, bool test_accept=false)
```
In commit "[validation] Add CChainState::ProcessTransaction()" (bbc785f31b5d70c702ae37d259c46652e243ddba)

Maybe for consistency with `ProcessNewBlock` and `ProcessNewBlockHeaders`, call this method `ProcessNewTransaction` instead of `ProcessTransaction`. Might also help suggest that purpose of processing is to add new data, not validate existing data.

jnewbery replied:
Haha you noticed. I actually hope that we'd go the other way and change `ProcessNewBlock` to `ProcessBlock` and `ProcessNewBlockHeaders` to `ProcessBlockHeaders`. There's nothing to prevent passing in the same block/headers/transaction that have previously been processed, and the functions have ways of returning that information to the caller (`ProcessNewBlock` will set the `new_block` out-param to false and `ProcessTransaction` will set the `TxValidationState` result to `TX_CONFLICT` to indicate "Tx already in mempool or conflicts with a tx in the chain"). I think the `New` in `ProcessNew` is therefore redundant and perhaps even a bit misleading.

jnewbery committed:
[logging/documentation] Remove reference to AcceptToMemoryPool from error string

User-facing error messages should not leak internal implementation
details like function names. Update the MEMPOOL_REJECTED error string
from "Transaction rejected by AcceptToMemoryPool" to the more generic
"Transaction rejected by mempool". Also update the MEMPOOL_ERROR error
message from "AcceptToMemoryPool failed" to the more precise "Mempool
internal error" since this error indicates and internal (e.g.
logic/hardware/etc) failure, and not a transaction rejection.

jnewbery commented on src/validation.cpp:
```diff
@@ -3423,6 +3423,19 @@ bool ChainstateManager::ProcessNewBlock(const CChainParams& chainparams, const s
+    CChainState& active_chainstate = ActiveChainstate();
+    if (!active_chainstate.m_mempool) {
+        TxValidationState state;
+        state.Invalid(TxValidationResult::TX_NO_MEMPOOL);
```
Good point. Fixed!

jnewbery commented on src/util/error.cpp:
```diff
@@ -20,9 +20,9 @@ bilingual_str TransactionErrorString(const TransactionError err)
             return Untranslated("Peer-to-peer functionality missing or disabled");
         case TransactionError::MEMPOOL_REJECTED:
-            return Untranslated("Transaction rejected by AcceptToMemoryPool");
+            return Untranslated("Transaction rejected by mempool");
         case TransactionError::MEMPOOL_ERROR:
-            return Untranslated("AcceptToMemoryPool failed");
+            return Untranslated("Mempool logic error");
```
That's a fair point. I've split this change into its own commit with the reasoning in the commit log.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/23173#discussion_r742027928

> I've split this change into its own commit with the reasoning in the commit log.

Not sure if it's intentional, but part of edd4b472adbe6f45aee40d739f6c1f44e2dc35e2 that was changing "transaction pool" to "mempool" in some comments was also dropped. Seem fine but just checking it wasn't lost accidentally.

jnewbery commented on src/test/txvalidation_tests.cpp:
```diff
@@ -38,7 +38,7 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_reject_coinbase, TestChain100Setup)
     unsigned int initialPoolSize = m_node.mempool->size();
     const MempoolAcceptResult result = AcceptToMemoryPool(m_node.chainman->ActiveChainstate(), *m_node.mempool, MakeTransactionRef(coinbaseTx),
-                true /* bypass_limits */);
+                false /* bypass_limits */);
```
That's correct. I've updated the commit log as suggested.

jnewbery commented:
Thanks for the thorough review @ryanofsky! I believe I've addressed all of your review comments.

stickies-v reviewed (commented):
2c64270 seems to indicate that all AcceptToMemoryPool() calls from outside validation.cpp/h are removed, but it appears there are still [two](https://github.com/jnewbery/bitcoin/blob/2c64270bbe523ef87e7225c351464e7c716f0b3e/src/test/fuzz/tx_pool.cpp#L233) [references](https://github.com/jnewbery/bitcoin/blob/2c64270bbe523ef87e7225c351464e7c716f0b3e/src/test/fuzz/tx_pool.cpp#L333) left. Is this intentional?

jnewbery commented:
> 2c64270 seems to indicate that all AcceptToMemoryPool() calls from outside validation.cpp/h are removed, but it appears there are still two references left. Is this intentional?

Yes, see the `AcceptToMemoryPool` doxygen comment:

```
 * Try to add a transaction to the mempool. This is an internal function and is
 * exposed only for testing. Client code should use ChainstateManager::ProcessTransaction()
```

Those two remaining calls to ATMP are the tests that are referenced in that comment. They can't use the `ChainstateManager::ProcessTransaction()` interface function since they're testing a mempool that isn't owned by a `CChainState`.

lsilva01 reviewed (approved):
tACK 0fdb619 on Ubuntu 20.04

This PR successfully removes calls to `AcceptToMemoryPool()` from the net_processing layer and allocates them to the validation layer (in the new `ProcessTransaction` method and in `ActivateBestChainStep`).

It also adds a `mempool.check()` call after processing a new transaction. It has no effect on the default setting as `m_check_ratio` will be 0 and `CTxMemPool ::check()` will return immediately.

`txvalidation_tests` and `txvalidationcache_tests` ran successfully after changing `bypass_limits` from `true` to `false`.

theStack reviewed (approved):
Code-review ACK 0fdb619aaf1d62598263361a6082d182be1af792

Nice improvement! Happy to see a more consistent higher-level interface for adding transactions to the mempool via `ChainstateManager`, with an automatic `CTxMemPool::check()` included (which has no effect on mainnet by default, i.e. possible performance losses don't affect users).

ryanofsky reviewed (approved):
Code review ACK 0fdb619aaf1d62598263361a6082d182be1af792. Only changes since last review: splitting & joining commits, adding more explanations to commit messages, tweaking MEMPOOL_ERROR string, fixing up argument name comments.

This might be ready for merge. It's also a pretty straightforward review if anyone else wants to take a look.
